src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/EnumeratingWhitelist.java

```diff
@@ -50,90 +50,106 @@
+        String key = Whitelist.canonicalConstructorSig(constructor);
+        Boolean b = permittedCache.get(key);
+        if (b != null) {
+            return b;
+        }
+
+        boolean output = false;
         for (NewSignature s : newSignatures()) {
             if (s.matches(constructor)) {
-                return true;
+                output = true;
+                break;
             }
         }
-        return false;
+        permittedCache.put(key, output);
+        return output;
     }
 
     @Override public final boolean permitsStaticMethod(Method method, Object[] args) {
+        String key = Whitelist.canonicalStaticMethodSig(method);
+        Boolean b = permittedCache.get(key);
+        if (b != null) {
+            return b;
+        }
+
+        boolean output = false;
         for (MethodSignature s : staticMethodSignatures()) {
             if (s.matches(method)) {
-                return true;
+                output = true;
+                break;
             }
         }
-        return false;
+        permittedCache.put(key, output);
+        return output;
     }
 
     @Override public final boolean permitsFieldGet(Field field, Object receiver) {
+        String key = Whitelist.canonicalFieldSig(field);
+        Boolean b = permittedCache.get(key);
+        if (b != null) {
+            return b;
+        }
+
+        boolean output = false;
         for (FieldSignature s : fieldSignatures()) {
             if (s.matches(field)) {
-                return true;
+                output = true;
+                break;
             }
         }
-        return false;
+        permittedCache.put(key, output);
+        return output;
     }
 
     @Override public final boolean permitsFieldSet(Field field, Object receiver, Object value) {
-        for (FieldSignature s : fieldSignatures()) {
-            if (s.matches(field)) {
-                return true;
-            }
-        }
-        return false;
+        return permitsFieldGet(field, receiver);
     }
 
     @Override public final boolean permitsStaticFieldGet(Field field) {
-        for (FieldSignature s : staticFieldSignatures()) {
-            if (s.matches(field)) {
-                return true;
-            }
+        String key = Whitelist.canonicalStaticFieldSig(field);
+        Boolean b = permittedCache.get(key);
+        if (b != null) {
+            return b;
         }
-        return false;
-    }
 
-    @Override public final boolean permitsStaticFieldSet(Field field, Object value) {
+        boolean output = false;
         for (FieldSignature s : staticFieldSignatures()) {
             if (s.matches(field)) {
-                return true;
+                output = true;
+                break;
             }
         }
-        return false;
+        permittedCache.put(key, output);
+        return output;
     }
 
-    public static @Nonnull String getName(@Nonnull Class<?> c) {
```

Comment:
Incompatible. You cannot move a `static` method to a “superclass”; only javac follows that lookup chain, not the JVM AFAIK.

Anyway `EnumeratingWhitelist` was the right place for these to begin with I think.

Reply:
Yeah, yeah -- I need to do a bit of cleanup on the PR before it's ready for full review.  That's why it's WIP - I wanted to spin out a few ideas before finalizing.